### DISCUSSION TITLE
What effects FDS RAM usage?

### DISCUSSION THREAD

**Discussion by norm07** [firemodels/fds#13434] [Q&A] [answered]
Hi All,

I've recently been running models where I've mucked around with sprinklers and its blown out my RAM usage. Mesh size is around 200,000. However the sprinklers are running at 339 l/s, constant particle diameter, particle count of 10,000, velocity of 18.1.

What settings in general (besides mesh) cause a dramatic increase in RAM usage? 

Kind regards,
N.

**Comment by drjfloyd** [comment]
Are you using the current release of FDS? If not try that. We keep tweaking how particle memory is managed (it needs special treatment since particles can move between meshes and hence between different MPI processes).

By particle count of 10,000 do you mean the number of particles per second per sprinker or the total number of particles? If it is the particle per second from the sprinkler, then you might just be accumulating a lot of particles in your domain. You could limit the particle age or maximum number of particles.

**Reply by norm07** [reply]
With respect to particles its the former. Would reducing the particle age to 2 to 3 seconds be suitable and realistic?

**Reply by drjfloyd** [reply]
If after 2 to 3 s, particles have done what you need them to do for your specific usage of FDS, then sure a 2 to 3 s age would be suitable.

**Reply by norm07** [reply]
I'm also constantly running into issues where the particles are causing drops in the neighbouring gas temperature such that I get  the warning message. This is at a particle count per second of 10,000. My sprinklers are running at 314 L/min into a 0.25x0.25x0.25m grid mesh with an offset of 0.3m. I'm not sure why I'm still running into all these particle errors.

**Reply by drjfloyd** [reply]
If you are not using the current release, try using it. Beyond that we would need to see your input file.

**Reply by norm07** [reply]
[202046_Lidcombe_T2_DF1_SEN.txt](https://github.com/user-attachments/files/17069367/202046_Lidcombe_T2_DF1_SEN.txt)

I tried running it on the latest release to no avail. Please see my model attached.

**Reply by norm07** [reply]
I should add that the RAM usage has come down, but I'm still getting the "WARNING! TEMP...etc" error.

**Reply by drjfloyd** [reply]
18.3.1 Sprinklers

INITIAL_TEMPERATURE of the link in units of ◦C. (Default TMPA)

Droplets default to an initial temperature of TMPA. Your TMPA is below the freezing point of water (&MISC TMPA=-20.0/) which is causing the warning. Set INITIAL_TEMPERATURE on PART.

**Reply by norm07** [reply]
My apologies, I sent you an older file. I have made the above change and still receive the warning message. This is with a particle count of 100,000. No idea whats going on...

**Reply by mcgratta** [reply]
Post a file that demonstrates the problem. Make sure that this file runs with the latest version of FDS.

**Reply by drjfloyd** [reply]
I set INITIAL_TEMPERATURE to 20 C and started the case with the current source. It is now 15 s beyond sprinkler operation. With AGE set to 2 this means droplets are at a quasi-steady state.  I am not seeing any WARNING messages so far.

**Reply by drjfloyd** [reply]
I got a warning at 191 s and the simlation seems to have hung. Will take some time to figure out what might be going on.

**Reply by norm07** [reply]
Hi Doc, I had an insufficient virtual memory warning. I've attached the file that I ran. (it ran for 265 seconds).
[202046_Lidcombe_T2_DF1_SEN.txt](https://github.com/user-attachments/files/17122076/202046_Lidcombe_T2_DF1_SEN.txt)

**Reply by norm07** [reply]
so i thought that maybe the particle count was driving up the ram usage so i reduced it by a factor of 10 for one of my other design fires (same compartment and sprinkler system etc), and i got the sprinkler particle warning this time round :/. File attached. 
[202046_Lidcombe_T2_DF1_WC.txt](https://github.com/user-attachments/files/17126072/202046_Lidcombe_T2_DF1_WC.txt)

**Reply by drjfloyd** [reply]
@mcgratta Been running this case with debug. Found a bug in how we handle particle WARNINGs that was setting up a infinite loop; however, the reason for the WARNING was Q_DOT_RAD in PARTICLE_MASS_ENERGY_TRANSFER going off the rails.  Right now we:

Move Drops
Evaporate drops using QR_W to get Q_DOT_RAD based on AVG_DROP_AREA from before moving.
Compute A_D_A after evaporation
Do radiation getting QR_W using the A_D_A after evaporation

Had a paricle cross mesh boundaries and it had a slight non-zero QR_W but an essentially zero AVG_DROP_AREA. As a result in Q_DOT_RAD = QR_W / A_D_A * ADROP was ~ -1 / 1E-117 *  * 1E-7 = -1 E110.  

1. We could evaporate before moving (i.e. move the call to PARTICLE_MASS_ENERGY_TRANSFER before MOVE_PARTICLES). Thought not sure if that would have some unforseen issue. This would make sure we evaporate with particles still where they were when A_D_A was last updated.
2.  We could put some limit on Q_DOT_RAD. Something limiting the ratio (ADROP/A_D_A) maybe: Q_DOT_RAD = QR_W * MIN(1._EB/WGT,A_DROP*0.25/AVG_ADROP_AREA)?

**Reply by mcgratta** [reply]
Yes, these cell-averaged values have always been troublesome. Let me take a look.

**Reply by mcgratta** [reply]
I like option 1. Limits and cut-offs always seem to foil us in the long run. I don't recall any reason for putting `MOVE_DROPLETS` ahead of `PARTICLE_MASS_ENERGY_TRANSFER`.  The first thing to do is to just move the call in `main.f90` and run the verification cases. That might bring to the surface some reason why we should not do it. I cannot think of one off the top of my head. 

Make this an issue and assign it to me. I can easily run the verification cases here in an hour or so.

**Reply by drjfloyd** [reply]
The only thing I can think of is fine watermist at operation where drops might evaporate in the first time step.  In that case, if evaporation was first then there would be no droplet drag when it came time to do movement. Unless we kept track of the old and new radius. Then we could impose drag as a time average. If we assume a constant evaporation rate then the old and new radius would give a function of radius over the timestep and we could average that radius squared over the timestep for drag.

**Reply by norm07** [reply]
thanks for looking into this guys. I will leave this in your capable hands for the time being. Just wondering, is there a work around for this at the moment?

**Reply by drjfloyd** [reply]
One issue was fixed today. If that passes firebot, then tomorrow there will be a test bundle at https://github.com/firemodels/test_bundles that you could try. That might help. If not then the other issue related to Q_DOT_RAD has a likely fix but that might not make it into tomorrow's bundle.

**Reply by norm07** [reply]
forgive my ignorance of the back end software stuff but are you saying i should wait for an updated FDS release to see these changes come through?

**Reply by drjfloyd** [reply]
Every night an automated tool called firebot checks out a fresh copy of the fds and smokeview git repositories, compiles fds and smv, runs a suite of ~900 test cases, checks all the results against known values, and builds the documentation. If all steps are successful, it then creates test bundles that are put on the github site in my prior post. You can look at the website tomorrow morning (Washington, DC time) and look to see if there is a new bundle for fds. If there is, you can download and install that bundle and try using it for your case. If that doesn't work then the next set of changes might but those are not comitted yet for firebot to test.

**Comment by norm07** [comment]
Hi Doc,

I've been having a great time modelling sprinklers following the update. Until I ran into some trouble with insufficient RAM.

Please see attached FDS file. I get the error at around 400 seconds (5th sprinkler activation from my view), do I simply just not have enough RAM or is this a similar bug as previous?

File attached.

Kind regards,
Norman.
[222150_DF1_SEN.txt](https://github.com/user-attachments/files/17389009/222150_DF1_SEN.txt)

**Reply by drjfloyd** [reply]
What is the error message you are getting?

**Reply by norm07** [reply]
insufficient virtual memory. In addition to the above model, I was also running a model of a similar size at the same time. My system has 32 Gb of memory.

**Reply by drjfloyd** [reply, marked as answer]
You may not be able to run two cases at once with your current inputs.  With 5 sprinklers operating and 10,000 particles per second per sprinkler, you are going to have a lot of particles to track.  If you are on a windows machine, depending on how much stuff you have running in the background you could be loosing 8-10 Gb just for the OS.  You coud try reducing the particles per second. You could make a case where you just start a sprinkler right away with TIME as the QUANTITY and look at how many particles per second you need to get a decent looking cascade of particles down the boxes.

**Reply by norm07** [reply]
Thanks for your input doc, ill muck around with it for a bit and see what works ... or spend some money on some RAM.
